### PR TITLE
[API][CLI] Add etcd-backed service discovery for standalone gateways

### PR DESCRIPTION
**Pull request by zupengwang**
## Pull Request Description

Standalone gateways can now discover inference endpoints from etcd without restarting or editing a static endpoint list. A registration is a JSON record under a configurable prefix; updates change the routing cache, and deletion or lease expiry withdraws the route.

- Add an etcd implementation of the existing discovery Provider: initial snapshot, revision-based watch, reconnect and compaction recovery, stable endpoint identity, and prefill/decode role metadata.
- Add strict `--etcd-config` loading with username/password and verified TLS/mTLS, plus configuration and registration examples.
- Make routing-cache registry snapshots safe during concurrent discovery updates. Real-etcd race tests exposed unsynchronized cache publication and reuse of slices already returned to requests; the registry now protects publication and creates a new snapshot after mutations.
- Build the gateway as a package so the new CLI file is included, and preserve bracketed IPv6 addresses in routing targets and prefill URLs.

## Related Issues

Related: #2033. This implements the etcd provider; Consul remains future work.

## Validation

Validated on base `5c77915f74d00a4dcdd6034615f0eba47b0af66f` using Go 1.22.12 and a dedicated etcd 3.5.17 server.

- `go test -race -p=1 -count=1 ./pkg/cache/... ./pkg/plugins/gateway/... ./pkg/types ./pkg/utils ./cmd/plugins`: 16 tested packages passed.
- `make lint-all`: license check and full golangci-lint passed.
- Real-etcd race integration: compaction recovery, routing-cache lifecycle, and natural lease expiry passed.
- Rebuilt the gateway with `CGO_ENABLED=0 go build -tags=nozmq ./cmd/plugins`.
- A real ext_proc gRPC client verified registration, address updates, invalid-update withdrawal, repair, deletion, and lease expiry through the gateway.
- A separate etcd server requiring client certificates verified gateway registration discovery and deletion watch over mTLS.

The blackbox validation covers ext_proc decisions, not Envoy forwarding or GPU inference. Multi-node etcd failure injection and throughput benchmarking were not run. During etcd outages, the provider retains the last known routes until discovery recovers, as documented.

### CODE REVIEW

**Comment by bolubo** [comment]
Hi @zupengwang, are you still planning to continue this PR? The last push was a couple of weeks ago and the branch now has conflicts with main. Asking before I look into the service discovery area (#2031, #2033) so we do not write it twice. If you are still on it, no problem at all.

**Comment by zupengwang** [comment]
@bolubo Yes, I’m continuing this PR. I’ll update it against current main and address the review feedback. Since #2728 has landed, I’ll reuse that Registry fix and keep this PR focused on etcd discovery through the existing Provider interface. Happy to coordinate on the broader #2031 work.

**Comment by bolubo** [comment]
Thanks, good to hear you are staying on it. I will leave the etcd provider to you and pick up the #2031 side: the static config surface first, then the label semantics, with the Endpoint type last.

If the abstraction turns out to need a change in `discovery.Provider` or `WatchEvent.Object` (`pkg/cache/discovery/discovery.go:53` and `:36`), I will say so on #2031 and link it here before opening anything, so you only have to rebase once. If you would rather get etcd in first and have the abstraction follow, that works for me too.

Once your update against main is up, I can take a look at the review items as well, if that would help. No rush either way, and good luck with the rebase.

File: pkg/cache/discovery/etcd.go [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+		stopWatch()
+		for {
+			if !waitEtcdRetry(ctx) {
+				return
+			}
+			if resync {
+				pods, snapshotRevision, err := p.snapshot(ctx, client)
+				if err != nil {
+					// Values, keys, endpoints, and credentials never enter this log.
+					klog.InfoS("Etcd discovery snapshot unavailable; retrying")
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/etcd.go]: Addressed in 2efd161. Watch and resnapshot retry logs now include a gRPC status code and a fixed diagnostic category for authentication, permission, TLS, timeout, transport, or other failures. Raw error text is not logged because it can contain endpoint or credential data. TestEtcdFailureDiagnosticsDoNotExposePayloads covers these categories with sensitive-looking payloads.

File: cmd/plugins/discovery.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+		// The etcd client selects one transport for all endpoints using the first URL.
+		if endpointScheme != "" && u.Scheme != endpointScheme {
+			return config, fmt.Errorf("etcd endpoints must use the same HTTP or HTTPS scheme")
+		}
+		endpointScheme = u.Scheme
+	}
+	config.Endpoints = file.Endpoints
+	config.Prefix = file.Prefix
+	if config.Prefix == "" {
+		config.Prefix = "/aibrix/endpoints/"
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply cmd/plugins/discovery.go]: Addressed in 2efd161. The CLI now delegates defaults and validation to NormalizeEtcdConfig, also called by NewEtcdProvider. It adds the trailing slash and rejects root and surrounding-whitespace prefixes. CLI and provider regression tests cover the same boundary cases.

File: cmd/plugins/discovery.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	if file.DialTimeout != "" {
+		timeout, err := time.ParseDuration(file.DialTimeout)
+		if err != nil || timeout <= 0 {
+			return config, fmt.Errorf("etcd dialTimeout must be a positive duration, such as 5s")
+		}
+		config.DialTimeout = timeout
+	}
+	config.Username = file.Username
+	config.Password = file.Password
+	if file.Password != "" && file.Username == "" {
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply cmd/plugins/discovery.go]: Addressed in 2efd161. Config-file credentials require HTTPS for every endpoint. NewEtcdProvider enforces encrypted transport too: native host:port endpoints require an explicit TLS config, and explicit HTTP is rejected even when TLS config is supplied. Tests cover HTTP rejection and HTTPS/TLS acceptance.

File: pkg/cache/discovery/etcd.go [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+	if err := reconcileEtcdPods(ctx, handler, state, pods); err != nil {
+		stopWatch()
+		cleanup()
+		return err
+	}
+	go func() {
+		defer cleanup()
+		p.run(ctx, client, handler, state, revision, stream, stopWatch)
+	}()
+	return nil
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/etcd.go:224]: Addressed using the documented snapshot-readiness option in 2efd161. The Watch contract and README now explicitly say readiness is at snapshot revision R, with R+1 catch-up and any compaction reconciliation asynchronous; registrations deleted after R may temporarily remain. The first compaction resnapshot also starts immediately, without the old one-second delay. Initial-delivery and real-etcd compaction tests pass.

File: pkg/utils/registry.go [hunk 5]
```diff
@@ -163,15 +170,15 @@ func (reg *CustomizedRegistry[V, A]) Len() int {
...
-	if !reg.valid && reg.registry != nil {
-		if cap(reg.values) < len(reg.registry) {
-			reg.values = make([]V, 0, len(reg.registry)*2)
-		}
-		for _, pod := range reg.registry {
-			reg.values = append(reg.values, pod)
+	if !reg.valid {
+		// A previous snapshot can still be in use by a routing request.
+		// Always allocate a new backing array when rebuilding the cache.
+		reg.values = make([]V, 0, len(reg.registry))
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/utils/registry.go]: Addressed in 2efd161. Main was merged, and the Registry implementation/tests now exactly match upstream #2728. The remaining PodArray issue is fixed separately: deployment groups no longer sort the published Pods slice, and sync.Once publishes the lazy indexes safely. TestPodArrayIndexPreservesPublishedSnapshot reproduced the old data race and now passes under -race, as do all 17 related packages.

File: pkg/cache/discovery/etcd.go [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+				// A watch-created response can have a header ahead of its events.
+				// Advance only past events actually applied, after the whole batch.
+				// The client coalesces fragmented responses by default.
+				revision = nextRevision
+				continue
+			}
+			if ctx.Err() != nil {
+				return
+			}
+			resync = response.CompactRevision > 0
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/etcd.go:279]: Took the diagnostic-logging alternative in 2efd161. Non-compaction disconnects still resume from the last applied revision + 1; compaction immediately triggers reconciliation. I did not add periodic resnapshots, since revision resume preserves the stream while history is available, and a snapshot would not repair an unchanged permission/auth failure. Those failures are now distinguishable in the retry logs.

File: pkg/types/router_context.go [hunk 7]
```diff
@@ -344,11 +346,11 @@ func (r *RoutingContext) targetAddress(pod *v1.Pod) string {
 		// back to the default deployment port on this warm pool pod.
 		return ""
 	}
-	return fmt.Sprintf("%v:%v", pod.Status.PodIP, utils.GetModelPortForPod(r.RequestID, pod))
+	return net.JoinHostPort(pod.Status.PodIP, strconv.FormatInt(utils.GetModelPortForPod(r.RequestID, pod), 10))
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/types/router_context.go:402]: Addressed in 2efd161. Both the engine scrape endpoint and the PromQL instance label now use net.JoinHostPort. TestMetricsWorkerScrapesIPv6 exercises the real worker HTTP path against an IPv6 loopback server; TestPromQLUsesIPv6InstanceLabel checks the generated queries. Both pass under -race. Routing and prefill IPv6 coverage was retained through the main merge.

File: pkg/cache/discovery/etcd.go [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+		return nil, fmt.Errorf("etcd discovery requires at least one etcd endpoint")
+	}
+	for _, endpoint := range config.Endpoints {
+		if strings.TrimSpace(endpoint) == "" {
+			return nil, fmt.Errorf("etcd discovery endpoints must not be empty")
+		}
+	}
+	if config.Prefix == "" {
+		config.Prefix = defaultEtcdPrefix
+	}
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/etcd.go:123]: Addressed in 2efd161. NormalizeEtcdConfig is shared by NewEtcdProvider and loadEtcdConfig, including the single default prefix, trailing-slash normalization, and rejection of root/blank/surrounding-whitespace prefixes. Both entry points have regression coverage.

File: pkg/cache/discovery/README.md [hunk 9]
```diff
@@ -60,6 +60,123 @@ models:
...
+| `address` | Required worker `host:port`; use `[IPv6]:port` for IPv6. Port must be 1–65535. |
+| `engine` | Optional engine label, such as `vllm`, `sglang`, or `trtllm`. |
+| `role` | Optional `prefill` or `decode`; requires `roleset`. |
+| `roleset` | Optional P/D pairing group; requires `role`. |
+
+Workers or an external controller own these registrations; the gateway only
+reads and watches them. Register each worker under a stable, unique key.
+Addresses must be reachable from the gateway. The bundled standalone Envoy uses
+`ORIGINAL_DST`, so register numeric IPv4 or IPv6 addresses when using that
+configuration; DNS names require a downstream proxy configured to resolve them.
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/README.md:125]: Addressed in 2efd161. The README now requires a read-only gateway identity scoped to the configured prefix, explains that trusted writers control routing destinations, and documents the snapshot-only readiness guarantee, asynchronous catch-up, and last-known-route behavior during outages.

File: pkg/cache/discovery/etcd.go [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+			if ctx.Err() != nil {
+				return
+			}
+			resync = response.CompactRevision > 0
+			klog.InfoS("Etcd discovery watch interrupted; reconnecting", "compacted", resync)
+		}
+
+		stopWatch()
+		for {
+			if !waitEtcdRetry(ctx) {
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/etcd.go]: Addressed in 2efd161. A compacted watch now performs its first resnapshot immediately; subsequent failed snapshots still back off. TestEtcdCompactionResnapshotStartsImmediately rejects the old mandatory one-second wait, and the real-etcd compaction recovery test passes. The asynchronous snapshot-readiness contract is also explicit in the Watch documentation and README.

File: pkg/cache/discovery/etcd.go [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+	}
+	host, portString, err := net.SplitHostPort(endpoint.Address)
+	if err != nil || host == "" {
+		return nil, fmt.Errorf("address must be host:port or [IPv6]:port")
+	}
+	if ip := net.ParseIP(host); ip != nil {
+		host = ip.String()
+	} else {
+		host = strings.ToLower(strings.TrimSuffix(host, "."))
+		if len(validation.IsDNS1123Subdomain(host)) != 0 {
```
[the remark this answers is not in the document]
Comment (zupengwang) [reply pkg/cache/discovery/etcd.go:442]: Addressed with the documented restriction option in 2efd161. The registration table and README explicitly require IP or DNS-1123 hostnames and reject underscores, noting the stricter behavior versus the legacy StaticProvider parser. The endpoint-validation table now includes an underscore-hostname rejection case; valid DNS names, IPv4, and IPv6 remain covered.

**Comment by zupengwang** [comment]
@bolubo The update is ready for another review.

I merged main at `8f44dea` in `200ba0f` and addressed the review feedback in `2efd161`. The PR now reuses #2728's Registry implementation; the remaining PodArray indexing race is fixed without mutating published snapshots. The follow-up also covers shared prefix/auth validation, immediate compaction recovery, safe diagnostics, IPv6 metrics, and the documented readiness/ACL/hostname contracts.

All 17 related packages passed `-race`, full lint/license checks passed, and the rebuilt gateway passed real-etcd integration, ext_proc lifecycle, and mTLS checks on the remote node.

The scope remains the etcd provider. Provider and WatchEvent.Object are unchanged, so your static configuration, label semantics, and Endpoint work for #2031 can continue independently. Could you take another look when you have time?

**Comment by bolubo** [comment]
Thanks for the writeup. I had another look at 2efd161, and what I could check on my side lines up with what you described, including the boundary: discovery.Provider, WatchEvent and the event types are untouched, and the TODO line in discovery.go is the only change in that file. So the #2031 side is free to move on its own; if the abstraction ever needs a change, I will bring it to #2031 first, as discussed.

I also ran the live etcd paths here against a throwaway etcd 3.5.17: compaction recovery (TestEtcdCompactionRecoveryIntegration), the routing-cache lifecycle, and lease expiry all passed. The discovery, types, utils, and pd/prefill packages passed under -race as well. I did not repeat the full -race sweep, the lint run, or the ext_proc and mTLS checks, so those still rest on your run. Nothing on my side blocks the update.

One small note from a wider sweep, out of scope for this PR: the gateway and cache paths now use net.JoinHostPort, but three controller-side sites still join the address by hand, which would not produce a valid address for an IPv6 pod IP: pkg/controller/modeladapter/utils.go:125 and :127, pkg/controller/modelclaim/runtime_client.go:217, and pkg/controller/podautoscaler/metrics/fetcher.go:94. No action needed in this PR; a follow-up issue would cover it if you agree it is worth tracking.
